docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+
+#: description="Set to deploy containerized node"
+r_openshift_node_is_containerized: false
+#: description="Set to deploy a node over AH"
+r_openshift_node_is_atomic: false
+```
+
+The role variables can be broken down intro three categories:
+* required role variables but undefined: role variables that must be always defined but without a default value
+* optional role variables but undefined: role variables that are undefined by default but can be defined
+* role variables that are always defined with a default value: role variables that can be overwritten
+
+All three categories can be annotated with the following keywords:
+* name: optional annotation used for role variables that are undefined, used when a table of role variables and pre-checks are generated
+* description: short and accurate description of a role variable
+* required: set to `false` if not specified, if set to `true` the role variable is required (forced by generated pre-checks tasks)
+* choices: a list of acceptable values of a role variables
+* required_when: a list of conjuncts under which a role variable must be defined
+
+A table of role variables and a sequence of pre-checks are generated from the specified annotations.
+
+All variables under `var/main.yml` are considered local variables.
+
+#### We should check all relevant role variables are properly set
+
+A list of tasks checking role variables can be generated based on the role variable annotations.
+Given a list of role variables and their properties (e.g. if a variable is required, a list of accepted values)
+can change over time, it is important to keep the checks in sync with the `default/main.yml`.
+The following checks can be generated:
+
+```yaml
+- name: "Fail if r_openshift_node_deployment_type is not defined"
+  fail:
+    msg: "r_openshift_node_deployment_type must be specified for this role"
+  when:
+  - r_openshift_node_deployment_type is not defined
+
+- name: Fail if invalid r_openshift_node_deployment_type provided
+  fail:
+    msg: "r_openshift_node_deployment_type can only be set to a single value from ['origin', 'openshift-enterprise']"
+  when:
+  - r_openshift_node_deployment_type is defined
+  - r_openshift_node_deployment_type not in ['origin', 'openshift-enterprise']
+```
+
+All the checks can be stored into `tasks/pre_checks.yml` file and included in the main task file.
+
+#### Propagation of values to role dependencies
+
+A role can depend on another role that has its own set of role variables. There are two ways how to propagate values to role dependencies. First approach requires a caller has a knowledge of all role dependency variables:
+
+```yaml
+roles:
+- role: etcd
+  r_etcd_client_port: 2379
+  r_etcd_peer_port: 2380
+  r_etcd_common_etcd_runtime: docker
```

Comment:
I'm going to agree with @mtnbikenc.  If the role needs to call other roles then the role probably needs to be considered for a refactor.  I'd like it if the orchestration of roles occurs at the playbook level and not through role dependencies.  If we had roles where a calls b and b calls c, then we need to pass data through a to arrive at c.  `a -> b -> c`.  This tightly couples these roles together and obfuscates the role calls with their variables.

I would think that the calling playbook could just do the following:
```
include_role:
  name: a
vars:
  r_a_foo: bar

include_role:
  name: b
vars:
  r_b_foo: czar

include_role:
  name: c
vars:
 r_c_foo: far
```

This is more verbose but now allows us to be a bit more modular.  This approach is very flexible and allows for the ability to call a role when I need to call the role in a tool box fashion.